app/src/main/java/org/schabi/newpipe/player/mediabrowser/MediaBrowserConnector.java

```diff
@@ -0,0 +1,381 @@
+                .setTitle(streamHistoryEntry.getStreamEntity().getTitle())
+                .setSubtitle(streamHistoryEntry.getStreamEntity().getUploader())
+                .setIconUri(Uri.parse(streamHistoryEntry.getStreamEntity().getThumbnailUrl()));
+
+        return new MediaItem(builder.build(), MediaItem.FLAG_PLAYABLE);
+    }
+
+    private AppDatabase getDatabase() {
+        if (database == null) {
+            database = NewPipeDatabase.getInstance(playerService);
+        }
+        return database;
+    }
+
+    private LocalPlaylistManager getPlaylistManager() {
+        if (localPlaylistManager == null) {
+            localPlaylistManager = new LocalPlaylistManager(getDatabase());
+        }
+        return localPlaylistManager;
+    }
+
+    @Nullable Disposable bookmarksNotificationsDisposable;
+
+    private void setupBookmarksNotifications() {
+        bookmarksNotificationsDisposable = getPlaylistManager().getPlaylists().subscribe(
+                playlistMetadataEntries -> playerService.notifyChildrenChanged(ID_BOOKMARKS));
+    }
+
+    private void disposeBookmarksNotifications() {
+        if (bookmarksNotificationsDisposable != null) {
+            bookmarksNotificationsDisposable.dispose();
+            bookmarksNotificationsDisposable = null;
+        }
+    }
+
+    // Suppress Sonar warning replace list collection by Stream.toList call, as this method is only
+    // available in Android API 34 and not currently available with desugaring
+    @SuppressWarnings("squid:S6204")
+    private Single<List<MediaItem>> populateBookmarks() {
+        final var playlists = getPlaylistManager().getPlaylists().firstOrError();
+        return playlists.map(playlist -> playlist.stream()
+                .map(this::createPlaylistMediaItem)
+                .collect(Collectors.toList()));
+    }
+
+    private Single<List<MediaItem>> populatePlaylist(final long playlistId) {
+        final var playlist = getPlaylistManager().getPlaylistStreams(playlistId).firstOrError();
+        return playlist.map(items -> {
+            final List<MediaItem> results = new ArrayList<>();
+            int index = 0;
+            for (final PlaylistStreamEntry item : items) {
+                results.add(createPlaylistStreamMediaItem(playlistId, item, index));
+                ++index;
+            }
+            return results;
+        });
+    }
+
+    private void playbackError(@StringRes final int resId, final int code) {
+        playerService.stopForImmediateReusing();
+        sessionConnector.setCustomErrorMessage(playerService.getString(resId), code);
+    }
+
+    private void playbackError(@NonNull final ErrorInfo errorInfo) {
+        playbackError(errorInfo.getMessageStringId(), PlaybackStateCompat.ERROR_CODE_APP_ERROR);
+    }
+
+    private Single<PlayQueue> extractPlayQueueFromMediaId(final String mediaId) {
+        final Uri mediaIdUri = Uri.parse(mediaId);
+        if (mediaIdUri == null) {
```

Comment:
This condition will never be true: https://github.com/AndroidSDKSources/android-sdk-sources-for-api-level-21/blob/master/android/net/Uri.java

Reply:
Thanks, I'll fix this.